Lava: includable in mapconfig, new config options, some refactor

saurtron opened the pull request:
### Work done

- Refactor lavaConfig.lua module into Spring.Lava so it can be easier to access (only available to LuaUI and LuaRules for now).
- Make lava config includable in mapconfig/lava.lua inside mapfiles.
  - Also now game provided config goes into [common/configs/LavaMaps/](https://github.com/beyond-all-reason/Beyond-All-Reason/tree/05a9bb47b3d2f03f7736b290d0436cd0a7201594/common/configs/LavaMaps) 
- Add more lava config options:
  - The options were already inside map_lava.lua as TODO notes and comments.
  - See at the bottom of this PR for more details.
  - damageFeatures: enable damage to features and control damage speed to features.
  - effectDamage: set custom ceg effect for damage.
  - effectBurst: set custom ceg effect for bursts (or disable).
  - effectBurstSounds: set custom sounds for bursts.
  - ambientSounds: set custom ambient sounds (or disable).
- Slight cleanup of the lava code.
  - Extracted shaders to shaders/GLSL/lava/.
  - No more lava globals from the lavaConfig.lua (now lava.lua) module.
  - Optimized most critical parts from map_lava.lua a bit (tried not to overdo it).
- (maybe) Fixes stronghold lava effect. - actually this needs clarification

### Test steps

- Try your favourite lava map with this PR and check it still works the same (I tried mostly ghenna since that has a lava cycle).

You can try some of the new config options by editing modules/lava.lua directly, look for the `if` relevant to your map.

Examples:
```LUA
elseif string.find(mapName, "ghenna") then
  damageFeatures = true -- features will be damaged by lava
  effectDamage = "commander-spawn" -- commander spawn effect when a unit is damaged
  effectBurst = "commander-spawn" -- commander-spawn effect randomly around lava
  effectBurstSounds = {{"commanderspawn", 20, 60}, {"beep6", 40, 70}}  -- random sounds for bursts, not these ones don't work well since they're not mono and thus can't be positioned in 3d
  ambientSounds = {{"commanderspawn", 20, 60}, {"beep6", 40, 70}} -- some random ambient sounds
```

### Discussion

#### Stronghold fix

- According to @Nikuksis it's supposed to look like I'm setting it since it was changed to acid with the map remake this year.
  - There were two config stances, and the one not being applied seems to be the one added later and also the good one according to the lobby pic. Actually, there seems to be some fighting over it's style:
    - The acid one was added by @Nikuksis here: https://github.com/beyond-all-reason/Beyond-All-Reason/commit/77e9e05ce58dc9ae00b00a8811d8713273439e61 in jun 2024
    - It was set to lava by @icexuick  here: https://github.com/beyond-all-reason/Beyond-All-Reason/commit/dc3a68d6e1ea0c539ee261696a4455fa6d97c8ce in nov 2022


#### Namespace

- Spring.Lava might be temporary, I'd like to introduce Spring.Map to access other common functionality like getWaterLevel
  - I actually started this because I wanted to homogenize determining map water level/status in different places since it's not being done 100% correct everywhere and there's quite a few of duplicated code as well. Spring.Lava.isLavaMap is one important part of determining water status.
  - Can later do Spring.Lava and Spring.Map, or put lava inside map as Spring.Map.lava or Spring.Map.Lava. Also would like to be able to Spring.Map.isLavaMap directly.
  - I moved the luarules/configs/lavaConfig.lua code to modules/lava.lua. I think it's better there, but this is breaking change since api users might be VFS.including the old path. It can be moved back if needed. Also if more map stuff gets added like I say above it can be better to move it into modules/map/lava.lua.
  - I could introduce an empty Map module in this PR to leave a more final api and paths.

#### Including in maps

- With lava now includable in maps, the map specific config can now be migrated from the game to the mapfiles and removed from the codebase.
- I can migrate the config to the mapfiles, but have no idea atm where the code for them lives atm :P
- It will be simple, since creators just need to copy the config from [common/configs/LavaMaps/](https://github.com/beyond-all-reason/Beyond-All-Reason/tree/05a9bb47b3d2f03f7736b290d0436cd0a7201594/common/configs/LavaMaps) to mapconfig/lava.lua.

#### Config options

- There are more config options that could be added, but for now I tried to do just the ones I saw at the map_lava.lua file with comments like "TODO make this configurable". I have no special feeling about the options, they can be removed or changed.
  - Could make more bursts per second, now the ratio is fixed.
  - Could have arrays of cegdamage and cegburst instead of just one to choose from (it can be later upgraded to allow just one element or an array if the need arises).

#### New shaders location
- I extracted shaders from the file since they make working on the lua stuff very annoying and also looks cleaner this way. Would like to hear from @Beherith to see if he likes where I placed them or even getting them out of the file. Don't want to step into other ppl work.

#### Others
- There was code in the file to destroy features, but it was using SetFeatureReclaim, that doesn't update the gui very well so not sure if that's the best way or there's something else I'm missing. I changed it to SetFeatureResources where I can actually set metal, energy and reclaimLeft so it all updates properly at gui tooltip.
- Not sure the cegdamage effects play very well with features, there might be something going on there, but likely it's a limitation of ceg or just me being paranoid.
- I noticed an exception for air units so flying units won't die in lava even if not flying. This may be some kind of optimization. It's a bit weird landed units won't die, but didn't touch this for now since it doesn't look so important.
- While I refactored and reorganized a bit, I tried not to overdo it and there's a few things specially in map_lava.lua that can still be improved or optimized. Generally I have modified the tightest loops or code around what I already had to modify anyways.

### Screenshots:

#### Before and after comparison:

(should look the same, both images taken at 1:50 in game time)

![lava-looks-same](https://github.com/user-attachments/assets/6a6be1d1-df82-46e6-847b-846b190b915b)

#### Destroyable features

![lava-damages-features](https://github.com/user-attachments/assets/c9ad47e0-542e-4279-9011-f6cc78025aa0)

#### Custom ceg damage effect

testing random cegs from around the codebase

![lava-custom-ceg-damage](https://github.com/user-attachments/assets/4a2d6e30-1cc2-4327-aae7-168cd13cae2b)

#### Custom ceg burst effect

testing random cegs from around the codebase

![lava-custom-ceg-bursts](https://github.com/user-attachments/assets/da5db7cd-e167-41da-894b-806a180529e1)

#### Stronghold fix

This is how it looks in each version.

![lava-stronghold](https://github.com/user-attachments/assets/a26cbc52-acb6-42c5-a5b0-31f054a0a73b)

### Detailed new options

 #### damageFeatures

Enable damage to features and control damage speed. This works by setting it to true or a float. true will actually set 0.1. The value is the proportion of resources from the feature destroyed per second. So 0.1 means it will get destroyed in 10 seconds.

Default: `false`

#### effectDamage

Set custom ceg effect for damage. This is the effect applied when a unit is damaged or destroyed.

Default: `"lavadamage"`

#### effectBurst

Set custom ceg effect for bursts (or disable). A burst appears somewhere in the map every now and then.

Can be set to false and then no bursts will appear.

Default: `"lavasplash"`

#### effectBurstSounds

Set custom sounds for bursts. An array of sounds with min and max volume, one of them will be randomly selected every time a burst happens.

Default: `{ {"lavaburst1", 80, 100}, {"lavaburst2", 80, 100} }`

#### ambientSounds


Set of custom ambient sounds. An array of sounds with min and max volume. One of them will appear somewhere in the map every now and then. Can be set to false to disable ambient sounds.

Default:
```
{ {"lavabubbleshort1", 25, 65},
   {"lavabubbleshort2", 25, 65},
   {"lavarumbleshort1", 20, 40},
   {"lavarumbleshort2", 20, 40},
   {"lavarumbleshort3", 20, 40} }
```

saurtron commented on modules/lava.lua:
```diff
@@ -0,0 +1,539 @@
+
+	tideRhym = { { 19, 0.3, 5*6000 } }
+
+elseif string.find(mapName, "incandescence") then
```
Will do that, also maybe make some changes for compatibility with z-k, I like the idea of loading the config from mapconfig/lava_config.lua instead of mapinfo.lua for example, that way the file we provide under whatever/Configs/LavaConf/MapXX.lua can be the same as can be included into the mapfile.

I see they did add some of the things I added, albeit with different names, also they have slightly different names for some of the vars (like lOSdarkness instead of losDarkness), not sure how worth it it is to improve map compatibility though, since there might be other areas where the two projects diverged already.

saurtron replied:
Ok, done. I did it a bit different than z-k to allow for config to be loaded without the version name, otherwise it would break with any small change by the map creator. I still allow loading with version too though, but provide all configs without version in the name. Also I prioritize map provided config unless the game sets overrideMap in its map specific lava config.

Besides that, regarding compatibility with z-k I think it can be a bit hopeless so I load the config from a different path than them: mapconfig/lava.lua, this way maps could have both configs to support both styles.

Anyways let me know if you think more direct compatibility is wanted, but tbh I really don't like some of the capitalization they chose, like lOSdarkness, uVscale, tideamplitude, tideperiod. It's just an esthetic thing, so if compatibility is preferred I could use the capitalization they chose and also change a few names. Still, the implementation is likely to diverge so even if we provide "compatible" configs it's likely the lava would be broken in subtle ways.

That's why I chose to read a different file.

p2004a commented:
How often is map lava config modified in comparison to the rest of map file?

I assume this feature also means that texture will need to be distributed inside every map file so that it's all self contained?

The one thing I care about is that map files will become only more and more pure structured data, no code, no logic.

saurtron commented:
> How often is map lava config modified in comparison to the rest of map file?

No idea, but giving the map makers options to include the lava config will likely help them in tweaking it more.

> I assume this feature also means that texture will need to be distributed inside every map file so that it's all self contained?

It can be distributed inside the map file if it's not available inside bar already, yes.

> The one thing I care about is that map files will become only more and more pure structured data, no code, no logic.

This is just some structured data so I guess it's ok but let me know if you have any specific concerns.

saurtron commented:
Moved all map specific configuration to [common/configs/LavaMaps](https://github.com/beyond-all-reason/Beyond-All-Reason/tree/05a9bb47b3d2f03f7736b290d0436cd0a7201594/common/configs/LavaMaps). I'm introducing `common/configs`, not sure if that's the best place, but I don't like luarules/configs/ since being available to LuaUI I don't think this is specific to luarules. I can move it there if you prefer.

While doing this I noticed the following:

- Stronghold had two config definitions, looks like the latter one is the good one (it's acid instead of lava, like the map pic), but the game was loading the first one due to the ifdef structure. I have added the latter instead so this may be fixing the Stronghold map.
- Pit of Azar is in the configs, but the map is not provided atm by BAR. Somebody added the config at some point, not sure what to make of that. I have added the config but it will likely need to be tweaked if it gets added. I checked and it works but looks a bit too tiled after the map edges.
- Both "Seths Ravine Remake" and "MoonQ20XR2" maps had configs with `mapLava = false` set, this would make the lavaConfig not parse it. I'm not sure if they're supposed to have lava or damageWater in some way, but I checked and doesn't seem to be the case.

saurtron commented:
> The one thing I care about is that map files will become only more and more pure structured data, no code, no logic.

Actually, thinking about this, then maybe the config should be provided in the map as json files or similar, instead of lua files?

Thing is I don't see mapfiles using json files atm, so that should likely be done as a separate project to remove all lua from maps. Also maybe the project wants to use a different format instead of json if it ever goes in that direction.

p2004a commented:
> No idea, but giving the map makers options to include the lava config will likely help them in tweaking it more.

Map files don't have incremental updates, you have do download fill file, so if lava config is updated often, we might not want to have it in map files. Looking at [the file history](https://github.com/beyond-all-reason/Beyond-All-Reason/commits/master/luarules/configs/lavaConfig.lua) it doesn't seem to change very often, at least not the per-map values.

Also

> I can migrate the config to the mapfiles, but have no idea atm where the code for them lives atm :P

It kinda doesn't. There aren't really repos for maps AFAIK, definitely not for majority of them. I don't know how mappers are managing their maps, but it's not like we can send a PR somewhere to change particular map. Mappers are directly uploading the map archives themselves.

> Pit of Azar is in the configs, but the map is not provided atm by BAR. Somebody added the config at some point, not sure what to make of that

The map was released but then reverted because there were issues with it.

> Actually, thinking about this, then maybe the config should be provided in the map as json files or similar, instead of lua files? (...)

Yes, but as you say that should be separate project, for now I care there is no logic, so the question is effectively "Is it possible to put this into JSON", if the answer is "no", then IMHO it should not live there.

---

Anyway, I think moving away from per-map configuration inside of the game is good. It is problematic for all sorts of reasons:
- potential custom maps
- once we switch to slower release process where there is stable and test version, it will block releasing of maps

The alternative is to put this information in the out-of-mapfile [maps metadata](https://github.com/beyond-all-reason/maps-metadata) so just like start pos suggestions are done (passed in via modoption), how we want ffa startpos to be (currently also in game repo), how startboxes are configured etc. IMHO it needs Beherith input. The more dynamic this data is, the better it fits into maps metadata, the more static it is per map version, the more it belongs in the map files.

Beherith commented:
Fuck I totally forgot about this one: https://github.com/beyond-all-reason/Beyond-All-Reason/commits/Lava-Rework/

Beherith commented:
Regarding map/game metadata

My smartest approach so far was done for map_grass, where the config load order, from mapinfo.lua is:
1. Load config from map
2. Override from game if game override is present

The new cegs, while welcome, look strange as white/purple, @icexuick  ?

saurtron commented:
> Fuck I totally forgot about this one: https://github.com/beyond-all-reason/Beyond-All-Reason/commits/Lava-Rework/

lol, I can take a look and merge any changes into a branch if this gets merged, if you want, shouldn't be a problem

Beherith commented:
> > Fuck I totally forgot about this one: https://github.com/beyond-all-reason/Beyond-All-Reason/commits/Lava-Rework/
> 
> lol, I can take a look and merge any changes into a branch if this gets merged, if you want, shouldn't be a problem

Nah, I merged the main part of it in the commits above, namely the reload and fps smoothing

WatchTheFort commented:
I think it would be better to consolidate all of a map's config code into a single file, with one file per map. We can create a dedicated directory e.g. `maps/`.

This would be better than splitting a single map's config across multiple files.

saurtron commented:
> I think it would be better to consolidate all of a map's config code into a single file, with one file per map. We can create a dedicated directory e.g. `maps/`.
> 
> This would be better than splitting a single map's config across multiple files.

Are there other configs around? Or should I just place the ones here under maps/ ?

Beherith commented:
For the record, anyone adding a file into bar thats called `config.lua` will be taken behind the chemical shed and executed.

Damgam commented:
I'm just gonna say this.

I don't like some of these config options being available. The only configurable stuff should be the numbers, not the mechanics. Stuff like "damageMode" shouldn't be an option.

saurtron commented:
> features should always be killed by lava. Why make that an option?

I just implemented those options that seemed already there in the code but commented with (TODO make this configurable), that one tbh I think was just a bad (commented) implementation, anyways I can remove any of the options you don't like.

sprunk commented:
Configurable probably means customparams

saurtron commented:
> I don't like some of these config options being available. The only configurable stuff should be the numbers, not the mechanics. Stuff like "damageMode" shouldn't be an option.

Done, removed the damageMode and damageMinHealth.

sprunk commented on luarules/gadgets/map_lava.lua:
```diff
@@ -25,13 +25,10 @@ if gadgetHandler:IsSyncedCode() then
-	local lavaDamageMode = lava.damageMode
+
+	-- damage is specified in health lost per second, damage is applied every 10 frames
+	local lavaDamage = lava.damage/3.0
```
Might want to do
```lua
local DAMAGE_RATE = 10 -- frames
local lavaDamage = lava.damage * (DAMAGE_RATE / Game.gameSpeed)
```
and then `if frame % DAMAGE_RATE == 0` somewhere below where the magic constant 10 currently lives

saurtron replied:
done
